### ISSUE TITLE
Non-items dropped in custom derive

### ISSUE THREAD

**Issue by antoyo** [rust-lang/rust#48038]
Hi.
It looks like when you generate non-items in a custom derive, they are silently dropped.
For instance, with the following custom derive:
```rust
#[proc_macro_derive(HelloWorld)]
pub fn hello_world(input: TokenStream) -> TokenStream {
    let ast = syn::parse(input).unwrap();
    let gen = impl_hello_world(&ast);
    gen.into()
}

fn impl_hello_world(ast: &syn::DeriveInput) -> quote::Tokens {
    let name = &ast.ident;
    quote! {{
        fn hello_world() {
            println!("Hello, World! My name is {}", stringify!(#name));
        }
        hello_world();
    }}
}
```
and the following code:
```rust
fn main() {
    #[derive(HelloWorld)]
    struct _Toto {
    }
}
```
nothing is printed on the screen.
So, is it normal that non-items are dropped?
Could they be generated as well if it makes sense (like in a struct that is defined within a function)?
If not, if would be nice to have a least a warning when this happen.
Thanks.

cc @jseyfried @alexcrichton @nikomatsakis

**Comment by nikomatsakis** [comment]
It does seem like a warning (or error!) would be appropriate. I don't believe that custom derive was intended to be able to generate non-item content, however.

**Comment by jseyfried** [comment]
Yeah, this should definitely be an error (or warning cycle, hopefully not...).

Since the macro invocation is an item, the macro expansion is parsed as items. It looks like we are silently ignoring an error from the parser, which is scary.

**Comment by petrochenkov** [comment]
Fortunately this issue doesn't affect `proc_macro` and `proc_macro_attribute`.

**Comment by fmease** [comment]
Triage: No change.

MCVE:

* `rustc a.rs --crate-type proc-macro --extern proc_macro`:
  ```rs
  #[proc_macro_derive(D)]
  pub fn d(_: proc_macro::TokenStream) -> proc_macro::TokenStream {
      "{ dbg!(); }".parse().unwrap()
  }
  ```
* `rustc b.rs -L. --extern a`:
  ```rs
  #[derive(a::D)]
  struct T;
  fn main() {}
  ```

**Comment by theemathas** [comment]
There was previously #87316 which would fix this, but was closed due to inactivity.
